### ISSUE TITLE
[ENH] implement __getstate__

### ISSUE THREAD

**Issue by adrinjalali** [cython/cython#6534]
### Is your feature request related to a problem? Please describe.

In my code, I would like to do nothing different.

### Describe the solution you'd like.

Right now `__getstate__` returns nothing, while `__setstate__` correctly sets values. I think it'd be nice for `__getstate__` to return what can be passed to `__setstate__`. This information is already provided by `__reduce__()[1][2]`, but I'd like to be able to simply do:

```py
obj = MyCythonObj(...)
obj_state = obj.__getstate__()
new_obj = type(obj).__new__(type(obj))
new_obj.__setstate__(obj_state)
```

This is going to help us easily support cython objects in https://github.com/skops-dev/skops, which persists / loads objects w/o using pickle.

**Comment by scoder** [comment]
> Right now `__getstate__` returns nothing

That's probably the [default implementation of Python 3.11+](https://docs.python.org/3/library/pickle.html#object.__getstate__) that strikes here. If you try it under Py3.10 or earlier, I don't think you'll find a `__getstate__` method.

**Comment by adrinjalali** [comment]
> Isn't `__reduce__` the more generic interface from the consumer's point of view? If you target `__reduce__`, then it should work with all objects that support the pickle protocol, whereas if you require `__getstate__`, then that's not necessarily the case, AFAIU.

Yes it is, but it also opens the door to easily creating malicious pickle objects. The point in `skops` is that we don't rely on objects' `__reduce__`. It means we can't pickle everything, but we already support pretty much the whole sklearn objects. For instance, for numpy, we rely on `numpy.save/load` instead of pickling them.
